snap-update-tools/check_snap_availability.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+"""
+This program checks whether snaps with specified characteristics are available in the snap store.
+
+The matrix of all combinations is created and the program peridically checks whether all combinations are available until all of them are found or the timeout is reached.
+The characteristics that can be specified are:
+  - (required and one) version that all the snaps must have
+  - (required and one) channel that all of the snaps must be in
+  - (required and 1 or more) names of the snaps
+  - (optional and 1 or more) architectures that all of the snaps must be available for
+
+For instance we want to check whether there are following checkbox-related snaps:
+ - checkbox22
+ - checkbox16
+ - checkbox
+all having a version of "2.9.2-dev5-123abcdef"
+for the architectures:
+    - amd64
+    - arm64
+
+which in reality means we want to check whether there are following snaps:
+    - checkbox22_2.9.2-dev5-123abcdef_amd64.snap
+    - checkbox22_2.9.2-dev5-123abcdef_arm64.snap
+    - checkbox16_2.9.2-dev5-123abcdef_amd64.snap
+    - checkbox16_2.9.2-dev5-123abcdef_arm64.snap
+    - checkbox_2.9.2-dev5-123abcdef_amd64.snap
+    - checkbox_2.9.2-dev5-123abcdef_arm64.snap
+
+So the invocation of this program would be:
+    The specs are specified with a yaml that looks like this:
+    required-snaps:
+        - name: checkbox22
+            channels:
+                - edge
+            architectures:
+                - amd64
+                - arm64
+
+"""
+
+import argparse
+import requests
+import time
+import yaml
+from dataclasses import dataclass
+
+
+# the dataclass for the concrete snap specification,
+# instance of this class represents one, concrete snap
+@dataclass
+class SnapSpec:
+    name: str
+    version: str
+    channel: str
+    arch: str
+
+    def __hash__(self) -> int:
+        # needed so we can this dataclass instances as keys
+        return hash((self.name, self.version, self.channel, self.arch))
```

Comment:
if we make this a namedtuple we don't need this, why is this a `dataclass`? it seems that we never edit these fields

Reply:
addressed in https://github.com/canonical/hwcert-jenkins-tools/commit/f5493d726e52fabdf733474266ef20a51fb931f9